### ISSUE TITLE
Incorrect AlltoAll behavior in jitted function

### ISSUE THREAD

**Issue by EiffL** [mpi4jax/mpi4jax#176]
First of all, thanks for this really great project! I am very much looking forward to using it for building a jax-based distributed cosmological N-body simulation (https://github.com/DifferentiableUniverseInitiative/JaxPM). 
I'm encountering a very strange issue though when trying to JIT functions, which I can boil down to the minimum working example I present bellow involving preforming an AlltoAll collective to transpose an array:

```python
from mpi4py import MPI
import jax
import jax.numpy as jnp
import mpi4jax

# Create communicators
world = MPI.COMM_WORLD
rank = world.Get_rank()
size = world.Get_size()

# Gobal size of 2D array
shape = [256, 256]

# Create a slice of the 2D array on each process
master_key = jax.random.PRNGKey(42)
key = jax.random.split(master_key, size)[rank]
local_slice = jax.random.normal(key, [shape[0]//size, shape[1]])

@jax.jit
def transpose(arr):
    # Split the last dimension in how many processes we have
    arr = arr.reshape([shape[0]//size, size, shape[1]//size])
    arr = arr.transpose([1, 0, 2]) # Move the number of slices to first dimension
    arr, token = mpi4jax.alltoall(arr, comm=world)
    # Merge the collected slices of the first dimension
    arr = arr.reshape([shape[0], shape[1]//size]) 
    arr = arr.transpose([1,0]) # Moves the unsliced dimension to the end
    return arr, token

# Perform the transpose
transposed_slice, token = transpose(local_slice)

# Let's make sure we have waited for all computations to be done
token = mpi4jax.barrier(token=token)

# Let's check that the results are OK
transposed_array, token = mpi4jax.allgather(transposed_slice, comm=world, token=token)
transposed_array = transposed_array.reshape(shape)
total_array, token  = mpi4jax.allgather(local_slice, comm=world, token=token)
total_array = total_array.reshape(shape)

print("On rank %d, here is the result: "%rank, 
      jnp.abs(total_array.transpose([1,0]) - transposed_array).mean())
```

If the `transpose` function is jitted, the output on a 4 process mesh is:
```bash
$ srun python test_mpi4jax.py 
On rank 0, here is the result:  1.1207631
On rank 1, here is the result:  1.1207631
On rank 3, here is the result:  1.1207631
On rank 2, here is the result:  1.1207631
```

If I remove the jitting on that function, I get the expected behavior showing that the transpose is correctly computed:
```bash
$ srun python test_mpi4jax.py 
On rank 0, here is the result:  0.0
On rank 1, here is the result:  0.0
On rank 2, here is the result:  0.0
On rank 3, here is the result:  0.0
```
This example was run using jax v0.3.23, and mpi4jax v0.3.10.post6, and I also tried to revert back to a jax version from this summer, without much difference. I only tried on GPU but it appears to be unaffected by the use of CUDA-aware MPI or not. But enabling debug mode I can see that the communications do occur as expected.

Very interestingly, I noticed that if I tweak the code in the following way:
```python
@jax.jit
def transpose(arr):
    arr = arr.reshape([shape[0]//size, size, shape[1]//size])
    arr = jnp.einsum('ij,xjy->ixy', jnp.eye(size), arr) # instead of arr.transpose([1, 0, 2])
    arr, token = mpi4jax.alltoall(arr, comm=world)
    return arr, token

arr, token = transpose(local_slice)
arr = arr.reshape([shape[0], shape[1]//size]) 
transposed_slice = arr.transpose([1,0]) 
```
i.e. replacing the first `jnp.tranpose` by an equivalent but more complex op involving an `einsum` gives me the correct results. So I don't think there is something intrinsically wrong with mpi4jax, but something weird going on with the input and output buffers of the collective op, for lax particular ops.

This makes me think that the physical memory layout of the buffer at the output of the transpose op doesn't actually correspond to its theoretical shape, which may be some optimization done by XLA to prevent moving the memory layout back and forth between operations.

I'm not sure how to debug this further, any help or insight would be much appreciated! I'll continue to poke around anyway because I really really want to use mpi4jax for my application ^^

**Comment by dionhaefner** [comment]
Can __not__ reproduce on CPU. Could you confirm that?

**Comment by EiffL** [comment]
Hum.... interesting, yep, confirmed, no problem on CPU.

**Comment by EiffL** [comment]
But on a different machine, I just checked that the same thing happens with only 2 GPUs, even with this combination of versions:
```
In [4]: jax.__version__
Out[4]: '0.3.4'

In [5]: mpi4jax.__version__
Out[5]: '0.3.10.post6'
```

**Comment by EiffL** [comment]
Hum reading this: https://www.tensorflow.org/xla/shapes
Looks like there aren't many degrees of freedom to encode the physical and logical shape of an array, essentially the memory layout can account for row-major, axis-major, and some amount of padding.

A wild guess would be that the transpose op returns either row or column major arrays, based on what is the cheapest in terms of moving memory around. For a pure transpose, it's only a matter of flipping that switch.

Looking quickly at the mpi4jax code, I don't see an obvious place where the layout of the arrays is checked, is this happening somewhere already?

**Comment by dionhaefner** [comment]
Layout is not checked, everything is assumed to be C contiguous I think. So that's probably the source of the error.

**Comment by EiffL** [comment]
Ok, so I tried this: https://github.com/EiffL/mpi4jax/commit/b4561de473924770b8d35d26105f38767cb72b45
Which I think should work, except... it crashes violently because XLA seems unhappy with non-array inputs/outputs when a layout is specified.
```
2022-11-08 20:15:46.537224: F external/org_tensorflow/tensorflow/compiler/xla/shape.h:163] Check failed: IsArray() element_type: TOKEN
```
which happens [here](https://github.com/tensorflow/tensorflow/blob/79370fcbaf0026f043d5f06b1820896fabf21b53/tensorflow/compiler/xla/shape.h#L162) at a point where apparently the code is trying to access the layout of this shape. 

I guess it makes sense, a token has no layout, but then I don't see how to specify the layouts for some of the inputs/outputs and not the other...

**Comment by dionhaefner** [comment]
Thanks for looking into this. Your code looks good to me, this is what I would have tried, too. I presume this is a bug in JAX / XLA - we encounter these regularly since we seem to be the only major users of the token mechanism.

Do you have a good standalone example that shows the problem when using tokens with `CustomCallWithLayout`? In this case we should raise the issue. I wouldn't bet on it getting fixed because the custom call interface is deprecated, but if it's an easy fix they might do it.

Sound like the long-term solution is to switch to the new interface. In the meantime we could add something like `arr = jnp.ascontiguousarray(arr)` to all handlers. Could you try whether that fixes it?

**Comment by EiffL** [comment]
Yep, going to try to build a MWE, and the `jnp.ascontiguousarray(arr)` is a good idea! I'll try that

**Comment by EiffL** [comment]
Haven't built the example yet, but a few observations:

- jnp.ascontiguousarray is not implemented, that would have been too easy ^^'

- I have also tested using the new MLIR approach to define the custom op, same result. Reading a bunch more source code, I think the problem is actually on the xla server side, which attempts to enforce the layout for all variables without a token check. That means it's buried deep inside the xla code and a fix would take a long time to percolate to a released Jax version.

- a trivial fix would be to use an integer as a token, is there a good reason not to do that? Alternatively, create a dummy identity op that just enforces the correct layout, and apply it to all inputs of collective operations

**Comment by dionhaefner** [comment]
> That means it's buried deep inside the xla code and a fix would take a long time to percolate to a released Jax version.

Not necessarily. Bug fixes are usually fast (next minor release), even within XLA - if the fix is simple.

Sounds like this is above our pay grade, let's just raise the issue and see what happens. I'm not keen on using workarounds that bypass the token mechanism.

No pressure with that MWE, if you're busy I can create one in the next couple of days.

**Comment by EiffL** [comment]
All right, how about this for a MWE :-) 
https://colab.research.google.com/drive/1qS3yE5Zav2L5XNzszWmtXJY0uaXxP0d4?usp=sharing

Let me know if that brings the point home sufficiently clearly. And be my guest if you want to open a JAX issue.

**Comment by EiffL** [comment]
and in the meantime #178 solves my initial problem :-)

**Comment by dionhaefner** [comment]
I created an upstream issue, let's see what comes out of this.

**Comment by PhilipVinc** [comment]
By the way, as all MPI operations are array-wide (therefore they don't care about row or column major ordering) why can't we simply make sure that the output of the MPI operation has the same layout as the input?

**Comment by dionhaefner** [comment]
Correct me if I'm wrong but I think there is no exposed way to manipulate or even retrieve buffer layouts (these are determined at the XLA level, for example when XLA fuses subsequent transposes). Hence the need for `layout` parameters in custom call wrappers that inform XLA on what kind of layout we expect on inputs and outputs. 

I don't know if there's an input that says "I don't care about the layout as long as input and output layouts are the same".

**Comment by EiffL** [comment]
Yes exactly. Even for operations like send and recv which absolutely don't care about the content of the array (it's different for all2all, because it needs to slice the input buffer correctly), if XLA can give the send operation either row or column major buffers, on the recv end, you have no idea how you should read your array. As Dion said, you don't have that information at the op level. So I think it's important to enforce a convention.

**Comment by PhilipVinc** [comment]
Indeed you're right that the slicing in `alltoall` cares about the ordering... But then you don't need to apply your decorator to all the primitives we export.

I'm unsure about the ordering: I just tried and you can access an object in the encoding phase that gives you the layout of the input
```python
def mpi_alltoall_xla_encode_cpu(c, x, token, comm):
    comm = unpack_hashable(comm)

    shape = c.GetShape(x)
    dtype = shape.element_type()
    dims = shape.dimensions()
    layout = shape.layout()

```

the problem is that I can't use that to create the output layout, as 
```python
    sh = xla_client.Shape.tuple_shape(
        [
            xla_client.Shape.array_shape(dtype, dims, layout),
            xla_client.Shape.token_shape(),
        ]
``` 
fails..

**Comment by dionhaefner** [comment]
Good point regarding ambiguity between `send` and `recv`. If `send` can send data with any layout there is no way to communicate to the receiving process how to read it. So yes we need the fixed convention.

**Comment by PhilipVinc** [comment]
Yeah ok, let's play it safe and keep the convention everywhere...

**Comment by dionhaefner** [comment]
Sounds like a fix is going to go live soon. As soon as it lands I will probably take the opportunity and migrate everything to `custom_call`.

**Comment by EiffL** [comment]
Sounds good!

**Comment by PhilipVinc** [comment]
Jax released a new jaxlib version 18 hours ago, is this fixed by that?

**Comment by dionhaefner** [comment]
I'm working on a PR.

**Comment by dionhaefner** [comment]
Sorry for being 18 hours late :wink:

**Comment by PhilipVinc** [comment]
Haha, I thought it would work natively without switching to CustomCallWiyLayout but I guess I was wrong

**Comment by dionhaefner** [comment]
Actually switching to MLIR's custom call so we can close #150 too :v:

**Comment by dionhaefner** [comment]
@EiffL or @PhilipVinc could you confirm that this is fixed on `main`?

**Comment by EiffL** [comment]
Works like a charm! Awesome thanks so much!

**Comment by PhilipVinc** [comment]
We should add a test, however
Il 27 nov 2022, 16:12 +0100, Francois Lanusse ***@***.***>, ha scritto:

**Comment by dionhaefner** [comment]
Done.
